### PR TITLE
Add OpenTelemetry Support for Trace Decorator in PromptFlow

### PR DESCRIPTION
**Pull request by liucheng-ms**
# Description

This PR introduces OpenTelemetry support for the trace decorator in PromptFlow. The `trace` decorator, which is applied to tool and openai APIs, now generates OpenTelemetry spans containing execution information. This enhancement allows for detailed tracking of function execution within the PromptFlow scope.

Key changes include:
- Generation of a span before the execution of the decorated function.
- Setting span attributes with execution information such as input and TraceType.
- Addition of the function's result as an output attribute to the span.

A test has been added to validate these changes.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_core/tracer.py [hunk 1]
```diff
@@ -192,6 +197,40 @@ def _create_trace_from_function_call(
...
+    if tracer is not None:
+        return tracer._node_name
+    return None
+
+
+def enrich_span_with_trace(span, trace):
+    span.set_attributes(
+        {
+            "framework": "promptflow",
+            "span_type": f"promptflow.{trace.type}",
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/tracer.py:211]: Let's remove the `promptflow.` prefix.

File: src/promptflow/promptflow/_core/tracer.py [hunk 2]
```diff
@@ -192,6 +197,40 @@ def _create_trace_from_function_call(
...
+            "function": trace.name,
+            "inputs": serialize_attribute(trace.inputs),
+            "node_name": get_node_name_from_context(),
+            "tool_version": "tool_version",  # TODO: Check how to pass the tool version in
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/tracer.py:215]: Remove this if we don't know it now.

File: src/promptflow/promptflow/_core/tracer.py [hunk 3]
```diff
@@ -192,6 +197,40 @@ def _create_trace_from_function_call(
...
 
 
+def get_node_name_from_context():
+    tracer = Tracer.active_instance()
+    if tracer is not None:
+        return tracer._node_name
+    return None
+
+
+def enrich_span_with_trace(span, trace):
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/tracer.py:207]: Let's add fail safe check around this method.

Comment (liucheng-ms) [reply src/promptflow/promptflow/_core/tracer.py:207]: Sure.

**Review by thy09 — APPROVED** [review]

**Review by zzn2 — APPROVED** [review]

**Pull request merged**